File: mission-to-mars/src/test/java/mars/RoverTest.java
package mars;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

public class RoverTest {

	ModuleMartien sut;

	@Test
	void initialisationAtterissage() {
		// Arrange
		sut = new Rover(Direction.NORD, new Position(4, 4, 0));

		// Assert
		assertTrue(new Position(4, 4, 0).equals(sut.getPosition()));
	}

	@Test
	void doitAvancer1CaseVersNordSiTourneVersLeNord() {
		// Arrange
		sut = new Rover(Direction.NORD, new Position(4, 4, 0));

		// Act
		sut.avancer();

		// Assert
		assertTrue(new Position(4, 5, 0).equals(sut.getPosition()));
	}

	@Test
	void doitAvancer1CaseVersSudSiTourneVersLeSud() {
		// Arrange
		sut = new Rover(Direction.SUD, new Position(4, 4, 0));

		// Act
		sut.avancer();

		// Assert
		assertEquals(new Position(4, 3, 0), sut.getPosition());
	}

	@Test
	void doitAvancer1CaseVersOuestSiTourneVersLOuest() {
		// Arrange
		sut = new Rover(Direction.OUEST, new Position(4, 4, 0));

		// Act
		sut.avancer();

		// Assert
		assertEquals(new Position(3, 4, 0), sut.getPosition());
	}

	@Test
	void doitAvancer1CaseVersEstSiTourneVersLEst() {
		// Arrange
		sut = new Rover(Direction.EST, new Position(4, 4, 0));

		// Act
		sut.avancer();

		// Assert
		assertEquals(new Position(5, 4, 0), sut.getPosition());
	}

	@Test
	void doitReculer1CaseVersNordSiTourneVersLeNord() {
		// Arrange
		sut = new Rover(Direction.NORD, new Position(4, 4, 0));

		// Act
		sut.reculer();

		// Assert
		assertTrue(new Position(4, 3, 0).equals(sut.getPosition()));
	}

	@Test
	void doitReculer1CaseVersSudSiTourneVersLeSud() {
		// Arrange
		sut = new Rover(Direction.SUD, new Position(4, 4, 0));

		// Act
		sut.reculer();

		// Assert
		assertEquals(new Position(4, 5, 0), sut.getPosition());
	}

	@Test
	void doitReculer1CaseVersOuestSiTourneVersLOuest() {
		// Arrange
		sut = new Rover(Direction.OUEST, new Position(4, 4, 0));

		// Act
		sut.reculer();

		// Assert
		assertEquals(new Position(5, 4, 0), sut.getPosition());
	}

	@Test
	void doitReculer1CaseVersEstSiTourneVersLEst() {
		// Arrange
		sut = new Rover(Direction.EST, new Position(4, 4, 0));

		// Act
		sut.reculer();

		// Assert
		assertEquals(new Position(3, 4, 0), sut.getPosition());
	}

	@Test
	void neDoitPasMonter() {
		// Arrange
		sut = new Rover(Direction.EST, new Position(4, 4, 0));

		// Act + Assert
		assertThrows(UnsupportedOperationException.class, () -> sut.monter());
	}

	@Test
	void neDoitPasDescendre() {
		// Arrange
		sut = new Rover(Direction.EST, new Position(4, 4, 3));

		// Act + Assert
		assertThrows(UnsupportedOperationException.class, () -> sut.descendre());
	}

}

File: typescript/tests/Rover.test.ts
import { Direction } from '../src/Direction'
import { ModuleMartien } from '../src/ModuleMartien'
import { Position } from '../src/Position'
import { Rover } from '../src/Rover'

describe('Rover', () => {
  it('Initialisation de l’atterissage', () => {
    // WHEN
    const rover: ModuleMartien = new Rover(Direction.NORD, new Position(4, 4, 0));

    // THEN
    expect(new Position(4, 4, 0)).toStrictEqual(rover.getPosition());
  })

  it('Doit avancer d’une case vers le nord si tourné vers le nord', () => {
    // GIVEN
    const rover: ModuleMartien = new Rover(Direction.NORD, new Position(4, 4, 0));

    // WHEN
    rover.avancer();

    // THEN
    expect(new Position(4, 5, 0)).toStrictEqual(rover.getPosition());
  })

  it('Doit avancer d’une case vers le sud si tourné vers le sud', () => {
    // GIVEN
    const rover: ModuleMartien = new Rover(Direction.SUD, new Position(4, 4, 0));

    // WHEN
    rover.avancer();

    // THEN
    expect(new Position(4, 3, 0)).toStrictEqual(rover.getPosition());
  })

  it('Doit avancer d’une case vers l’ouest si tourné vers l’ouest', () => {
    // GIVEN
    const rover: ModuleMartien = new Rover(Direction.OUEST, new Position(4, 4, 0));

    // WHEN
    rover.avancer();

    // THEN
    expect(new Position(3, 4, 0)).toStrictEqual(rover.getPosition());
  })

  it('Doit avancer d’une case vers l’est si tourné vers l’est', () => {
    // GIVEN
    const rover: ModuleMartien = new Rover(Direction.EST, new Position(4, 4, 0));

    // WHEN
    rover.avancer();

    // THEN
    expect(new Position(5, 4, 0)).toStrictEqual(rover.getPosition());
  })

  it('Doit reculer d’une case vers le sud si tourné vers le nord', () => {
    // GIVEN
    const rover: ModuleMartien = new Rover(Direction.NORD, new Position(4, 4, 0));

    // WHEN
    rover.reculer();

    // THEN
    expect(new Position(4, 3, 0)).toStrictEqual(rover.getPosition());
  })

  it('Doit reculer d’une case vers le nord si tourné vers le sud', () => {
    // GIVEN
    const rover: ModuleMartien = new Rover(Direction.SUD, new Position(4, 4, 0));

    // WHEN
    rover.reculer();

    // THEN
    expect(new Position(4, 5, 0)).toStrictEqual(rover.getPosition());
  })

  it('Doit reculer d’une case vers l’est si tourné vers l’ouest', () => {
    // GIVEN
    const rover: ModuleMartien = new Rover(Direction.OUEST, new Position(4, 4, 0));

    // WHEN
    rover.reculer();

    // THEN
    expect(new Position(5, 4, 0)).toStrictEqual(rover.getPosition());
  })

  it('Doit reculer d’une case vers l’ouest si tourné vers l’est', () => {
    // GIVEN
    const rover: ModuleMartien = new Rover(Direction.EST, new Position(4, 4, 0));

    // WHEN
    rover.reculer();

    // THEN
    expect(new Position(3, 4, 0)).toStrictEqual(rover.getPosition());
  })

  it('Ne doit pas monter', () => {
    // GIVEN
    const rover: ModuleMartien = new Rover(Direction.NORD, new Position(4, 4, 0));

    // WHEN
    const monter = () => rover.monter()

    // THEN
    expect(monter).toThrowError('Un rover ne vole pas')
  })

  it('Ne doit pas descendre', () => {
    // GIVEN
    const rover: ModuleMartien = new Rover(Direction.NORD, new Position(4, 4, 0));

    // WHEN
    const descendre = () => rover.descendre()

    // THEN
    expect(descendre).toThrowError('Un rover ne vole pas')
  })
})
